clang/test/Parser/cxx2x-ambig-reflect-expr.cpp

```diff
@@ -0,0 +1,37 @@
+//===----------------------------------------------------------------------===//
+//
+// Copyright 2024 Bloomberg Finance L.P.
+//
+// Part of the LLVM Project, under the Apache License v2.0 with LLVM Exceptions.
+// See https://llvm.org/LICENSE.txt for license information.
+// SPDX-License-Identifier: Apache-2.0 WITH LLVM-exception
+//
+//===----------------------------------------------------------------------===//
+//
+// RUN: %clang_cc1 -std=c++26 -freflection -Wno-parentheses -verify %s
+
+
+enum MyEnum { x, y, e = -1, f, z = 99 };
+
+void func(MyEnum && x) { // ok
+  MyEnum value{};
+  MyEnum& ref = value;
+
+  constexpr auto reflValue = ^value;
+
+  constexpr bool test_comparison_0 = reflValue != (^MyEnum) && true; // ok
+  constexpr bool test_comparison_1 = (reflValue != (^MyEnum) && true); // ok
+
+  constexpr bool test_comparison_2 = ^MyEnum != (^MyEnum) && true; // ok
+  constexpr bool test_comparison_3 = (^MyEnum != (^MyEnum) && true); // ok
+
+  constexpr bool test_comparison_4 = (reflValue != (^MyEnum) & true); // ok
+
+  constexpr bool test_comparison_5 = reflValue != ^MyEnum && true;
+  // expected-warning@30 {{given expression may be misinterpreted due to 'T &&' type parsing logic. did you mean '... (^T) && ...'?}}
+  // expected-error@30 {{expected ';' at end of declaration}}
+  constexpr bool test_comparison_6 = reflValue != ^MyEnum & true;
+  // expected-warning@33 {{given expression may be misinterpreted due to 'T &' type parsing logic. did you mean '... (^T) & ...'?}}
+  // expected-error@33 {{expected ';' at end of declaration}}
+}
```

Comment:
I think there's an important missing case here, which is that e.g.,:
```cpp
constexpr auto r1 = ^int &;
constexpr auto r2 = ^int &&;
```

should both be ok without warning.

Reply:
Nice catch, missed it!

Fixed [a736536](https://github.com/bloomberg/clang-p2996/pull/55/commits/a7365364da1a4a782c0ed47e4d30b99e6f101d1c)